age-plugin.md

```diff
@@ -270,6 +280,32 @@ The following commands and responses are defined for this phase:
     be mapping the stanzas to specific recipients, and will cache stanzas for
     separate files until the state machine completes.
   - Response is `(ok)`.
+- `(labels, [LABEL [LABEL...]])` - if the client sent a `extension-labels` command
+  in phase 1, the plugin sends the set of labels it associates with the set of
+  recipient stanzas it generates in this phase.
+  - Labels can have any value, but usually take one of several forms:
+    - *Common public label* - used by multiple plugins to permit their recipient
+      stanzas to be used only together. Examples include:
+      - `postquantum` - indicates that the recipient stanzas being generated are
+        postquantum-secure, and that they can only be combined with other stanzas
+        that are also postquantum-secure.
+    - *Common private label* - used by plugins created by the same private entity
+      to permit their recipient stanzas to be used only together. For example, private
+      plugins used in a corporate environment could all send the same private label
+      in order to prevent compliant age clients from simultaneously wrapping file
+      keys with other plugins.
+    - *Random label* - used by plugins that want to ensure their stanzas are not
+      used with any other recipient stanzas. This can be used to produce a file key
+      that is only encrypted to a single recipient stanza.
+  - Plugins MUST NOT send duplicate labels, and MUST NOT send a `labels` command
+    more than once.
+  - The order of labels in the command is irrelevant. Clients MUST treat them as an
+    unordered set.
+  - Response is `(ok)` or an error.
```

Comment:
@FiloSottile you suggested the "or an error" here. What exactly did you intend to be returned to the plugin? In the rest of the plugin protocol we only define `(ok)` and `(fail)`, and the latter is _not_ an error resulting in a protocol abort. Did you mean that the client should abort the protocol and return an error _to the user_ if a MUST / MUST NOT is violated? Or did you intend to convey an error to the plugin?

Reply:
I agree with dropping the error condition (reverting your suggestion that I merged before noticing this issue).

I think the next bullet point was about avoiding the more intensive plugin operations if they were going to be pointless because the client was about to abort the entire encryption process. Labels are a bit different than user interactions in that it is unambiguous to a label-aware client that the subsequent plugin operations are going to be pointless. But the client dropping the connection on its own isn't necessarily going to halt them; only the client closing the plugin subprocess will do so, and that won't necessarily actually halt the operations (if the plugin called out to a long-running service and was just waiting on a response). And in my client plugin implementation I also didn't have a great way to halt the process early, indeed due to designing my APIs to enforce the plugin being the controller of that phase. So yeah, I think dropping the next bullet point would also be fine.

Reply:
I went to implement this and had a look at the commit history. Turns out we did originally allow returning `(fail)` from here if certain `labels` invariants weren't upheld, but replaced that with `MUST NOT` language in f2c4053fa4921ae72b89be45e925b187807fedcd.

Something we _could_ therefore do is return `(fail)` if the client knows that the plugin stanzas will not be compatible, but not _require_ that the client inform the plugin at this stage. That would be compatible with `WrapWithLabels` (the client would not ever tell the plugin what the existing label set it is working to is, just that it's incompatible with the plugin's label set, so `WrapWithLabels` would be used with a client impl that never returns `(fail)`), while clients that want to short-circuit could go around that abstraction (by not implementing plugins on the client side using the `Recipient` interface).

However, this would also be a backwards-compatible change: existing clients would never return `(fail)` to new plugins, and new clients returning `(fail)` would cause an error with existing plugins which is what we want to happen (it would maybe just be an unexpected error message). So I think on balance it's fine to drop `(fail)`.

Reply:
@FiloSottile I've pushed a commit making the change, with an extensive comment. Please review the comment and resolve this conversation if you're satisfied.

Reply:
Yep, thank you for documenting this.